consensus/types.go

```diff
@@ -194,30 +219,79 @@ func (b Block) MerkleRoot() hash.Hash {
 
 // MinerPayoutID returns the ID of the payout at the given index.
 func (b Block) MinerPayoutID(i int) OutputID {
-	return OutputID(hash.HashBytes(encoding.MarshalAll(b.ID(), i)))
+	return OutputID(hash.HashAll(b.ID(), i))
 }
 
-// FileContractID returns the id of a file contract given the index of the contract.
-//
-// TODO: Reconsider how file contract ids are derived
+// FileContractID returns the id of a file contract given the index of the
+// contract. The id is derived by marshalling all of the fields in the
+// transaction except for the signatures and then appending the string "file
+// contract" and the index of the contract.
 func (t Transaction) FileContractID(i int) ContractID {
-	return ContractID(hash.HashBytes(encoding.MarshalAll(
-		t.Outputs[0],
-		t.FileContracts[i],
-		"contract",
+	return ContractID(hash.HashAll(
+		FileContractIdentifier,
+		t.Inputs,
+		t.MinerFees,
+		t.Outputs,
+		t.FileContracts,
+		t.StorageProofs,
+		t.SiafundInputs,
+		t.SiafundOutputs,
+		t.ArbitraryData,
 		i,
-	)))
+	))
 }
 
-// OuptutID takes the index of the output and returns the output's ID.
-//
-// TODO: ID should not include the signatures.
+// OutputID gets the id of an output in the transaction, which is derived from
+// marshalling all of the fields in the transaction except for the signatures
+// and then appending the string "siacoin output" and the index of the output.
 func (t Transaction) OutputID(i int) OutputID {
-	return OutputID(hash.HashBytes(encoding.MarshalAll(
-		t,
-		"coinsend",
+	return OutputID(hash.HashAll(
+		SiacoinOutputIdentifier,
+		t.Inputs,
+		t.MinerFees,
+		t.Outputs,
+		t.FileContracts,
+		t.StorageProofs,
+		t.SiafundInputs,
+		t.SiafundOutputs,
+		t.ArbitraryData,
+		i,
+	))
+}
+
+// StorageProofOutputID returns the OutputID of the output created during the
+// window index that was active at height 'height'.
+func (fcID ContractID) StorageProofOutputID(proofValid bool) (outputID OutputID) {
+	outputID = OutputID(hash.HashAll(
+		fcID,
+		proofValid,
+	))
+	return
+}
+
+// SiafundOutputID returns the id of the siafund output that was specified and
+// index `i` in the transaction.
+func (t Transaction) SiafundOutputID(i int) OutputID {
+	return OutputID(hash.HashAll(
+		SiafundOutputIdentifier,
+		t.Inputs,
+		t.MinerFees,
+		t.Outputs,
+		t.FileContracts,
+		t.StorageProofs,
+		t.SiafundInputs,
+		t.SiafundOutputs,
+		t.ArbitraryData,
 		i,
-	)))
+	))
+}
+
+// SiaClaimOutputID returns the id of the siacoin output that is created when
+// the siafund output gets spent.
+func (id OutputID) SiaClaimOutputID() OutputID {
+	return OutputID(hash.HashAll(
+		id,
+	))
```

Comment:
are more fields going to be added to this? If not, you can just use `hash.HashObject`